cmd/flexvolume/cmd/mount.go

```diff
@@ -0,0 +1,149 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/rook/rook/pkg/agent/flexvolume"
+	"github.com/spf13/cobra"
+)
+
+var (
+	mountCmd = &cobra.Command{
+		Use:   "mount",
+		Short: "Mounts the volume to the pod volume",
+		RunE:  mount,
+	}
+)
+
+func init() {
+	RootCmd.AddCommand(mountCmd)
+}
+
+func mount(cmd *cobra.Command, args []string) error {
+
+	client, err := getRPCClient()
+	if err != nil {
+		return fmt.Errorf("Rook: Error getting RPC client: %v", err)
+	}
+
+	var opts = &flexvolume.AttachOptions{}
+	if err := json.Unmarshal([]byte(args[1]), opts); err != nil {
+		return fmt.Errorf("Rook: Could not parse options for mounting %s. Got %v", args[1], err)
+	}
+	opts.MountDir = args[0]
+
+	err = client.Call("FlexvolumeController.GetAttachInfoFromMountDir", opts.MountDir, &opts)
+	if err != nil {
+		log(client, fmt.Sprintf("Attach volume %s/%s failed: %v", opts.Pool, opts.Image, err), true)
+		return fmt.Errorf("Rook: Mount volume failed: %v", err)
+	}
+
+	// Attach volume to node
+	log(client, fmt.Sprintf("calling agent to attach volume %s/%s", opts.Pool, opts.Image), false)
+	var devicePath string
+	err = client.Call("FlexvolumeController.Attach", opts, &devicePath)
+	if err != nil {
+		log(client, fmt.Sprintf("Attach volume %s/%s failed: %v", opts.Pool, opts.Image, err), true)
+		return fmt.Errorf("Rook: Mount volume failed: %v", err)
+	}
+
+	mounter := getMounter()
+
+	// Mount the volume to a global volume path
+	var globalVolumeMountPath string
+	err = client.Call("FlexvolumeController.GetGlobalMountPath", opts.VolumeName, &globalVolumeMountPath)
+	if err != nil {
+		log(client, fmt.Sprintf("Attach volume %s/%s failed. Cannot get global volume mount path: %v", opts.Pool, opts.Image, err), true)
+		return fmt.Errorf("Rook: Mount volume failed. Cannot get global volume mount path: %v", err)
+	}
+	notMnt, err := mounter.Interface.IsLikelyNotMountPoint(globalVolumeMountPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(globalVolumeMountPath, 0750); err != nil {
+				return fmt.Errorf("Rook: Mount volume failed. Cannot create global volume mount path dir: %v", err)
+			}
+			notMnt = true
+		} else {
+			return fmt.Errorf("Rook: Mount volume failed. Error checking if %s is a mount point: %v", globalVolumeMountPath, err)
+		}
+	}
+	options := []string{opts.RW}
+	if notMnt {
+		err = redirectStdout(
+			client,
+			func() error {
+				if err = mounter.FormatAndMount(devicePath, globalVolumeMountPath, opts.FsType, options); err != nil {
```

Comment:
I'm also struggling with why the format and mount happens here at all. Shouldn't this logic all go inside the rook-agent and this function merely just call the rook-agent to do all of this?

Reply:
Reason is that the agents dont know ahead of time where kubernetes is expecting the mountDir to be. And since the rook-agent runs in a pod, it needs to have the mountDir host mounted ahead of time so that it can perform host-level operation like mount and format. We could do a "best guess" and have the agent pod host-mount `{kubelet-root}/pods/`. The `kubelet-root` folder will be derived from the operator pod. So this means that all nodes should be consistently configured.

But this would make the driver dumb and have all the logic in the agents, which is what i think you advocating to.

Reply:
I am very much pushing for a dumb driver -- it should only change if the flex API changes. All other changes like supporting nbd-rbd or support for cephfs and others should not require any changes in flex. I see the problem with mount path, and I wonder if there is a reasonable way around it.

Reply:
I could assume that the mountDir will be in the `{kubelet-root}/pods` dir, which is currently the default. But if this changes, we would have to reach fast.